src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.cs

```diff
@@ -2,52 +2,48 @@
 
 using System.Collections.Generic;
 using System.Threading.Tasks;
-using Microsoft.Build.Construction;
 using Microsoft.VisualStudio.ProjectSystem.Properties;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties
 {
     internal abstract partial class AbstractBuildEventValueProvider : InterceptingPropertyValueProviderBase
     {
-        private readonly IProjectLockService _projectLockService;
+        private readonly IProjectAccessor _projectAccessor;
         private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IProjectThreadingService _threadingService;
         private readonly AbstractBuildEventHelper _helper;
 
         protected AbstractBuildEventValueProvider(
-            IProjectLockService projectLockService,
+            IProjectAccessor projectAccessor,
             UnconfiguredProject project,
+            IProjectThreadingService threadingService,
             AbstractBuildEventHelper helper)
         {
-            _projectLockService = projectLockService;
+            _projectAccessor = projectAccessor;
             _unconfiguredProject = project;
+            _threadingService = threadingService;
             _helper = helper;
         }
 
         public override Task<string> OnGetEvaluatedPropertyValueAsync(
             string evaluatedPropertyValue,
             IProjectProperties defaultProperties)
         {
-#pragma warning disable RS0030 // symbol IProjectLockService is banned
-            return _projectLockService.ReadLockAsync(async access =>
+            return _projectAccessor.OpenProjectXmlForReadAsync(_unconfiguredProject, projectXml =>
             {
-                ProjectRootElement projectXml = await access.GetProjectXmlAsync(_unconfiguredProject.FullPath);
-                return await _helper.GetPropertyAsync(projectXml, defaultProperties);
+                return _threadingService.ExecuteSynchronously(() => _helper.GetPropertyAsync(projectXml, defaultProperties));
             });
-#pragma warning restore RS0030 // symbol IProjectLockService is banned
         }
 
         public override async Task<string> OnSetPropertyValueAsync(
             string unevaluatedPropertyValue,
             IProjectProperties defaultProperties,
             IReadOnlyDictionary<string, string> dimensionalConditions = null)
         {
-#pragma warning disable RS0030 // symbol IProjectLockService is banned
-            await _projectLockService.WriteLockAsync(async access =>
+            await _projectAccessor.OpenProjectXmlForWriteAsync(_unconfiguredProject, projectXml =>
             {
-                ProjectRootElement projectXml = await access.GetProjectXmlAsync(_unconfiguredProject.FullPath);
-                await _helper.SetPropertyAsync(unevaluatedPropertyValue, defaultProperties, projectXml);
+                _threadingService.ExecuteSynchronously(() => _helper.SetPropertyAsync(unevaluatedPropertyValue, defaultProperties, projectXml));
```

Comment:
We should avoid blocking a thread pool thread for this, we should separate the async path from the reading/updating path, or look into deeper why this is async.

Reply:
It comes from reading the project properties
https://github.com/dotnet/project-system/blob/81492404fee8a9e9a53f07ed68958836e3d1a0f0/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs#L39

Would love to change this to be better. What should we do here instead?

Reply:
Was just thinking the same.

Also, what's the difference between `ExecuteSynchronously` and `JTF.Run`?

The instance above looks different. The return value from `ExecuteSynchronously` is returned into `OpenProjectXmlForReadAsync`.

Reply:
Lets see. I can pull the async call to `IProjectProperties defaultProperties` up so that only synchronus code runs in inside the lock.  Let me try that

Reply:
> Also, what's the difference between ExecuteSynchronously and JTF.Run?

[Nothing](https://devdiv.visualstudio.com/DevDiv/_git/CPS?path=%2Fsrc%2FMicrosoft.VisualStudio.ProjectSystem.Implementation%2FProjectServices%2FProjectMultiThreadedService.cs&version=GBmaster&line=144&lineStyle=plain&lineEnd=149&lineStartColumn=1&lineEndColumn=1)

Reply:
alright, this should be fixed now